baseplate/observers/prometheus.py

```diff
@@ -121,15 +127,98 @@ def on_child_span_created(self, span: Span) -> None:
         else:
             observer = PrometheusClientSpanObserver()
 
-        observer.on_set_tag("protocol", self.protocol)
         span.register(observer)
 
 
 class PrometheusClientSpanObserver(SpanObserver):
     def __init__(self) -> None:
-        logger.debug("PrometheusClientSpanObserver not implemented")
+        self.tags: Dict[str, Any] = {}
+        self.start_time: Optional[int] = None
+        self.metrics: Optional[
+            Union[PrometheusHTTPClientMetrics]
+        ] = None  # Add PrometheusThriftClientMetrics when implemented
+
+    def on_set_tag(self, key: str, value: Any) -> None:
+        self.tags[key] = value
+
+    def get_prefix(self) -> str:
+        return f"{self.protocol}_client"
+
+    def get_labels(self) -> Dict[str, str]:
+        return self.tags
+
+    @property
+    def protocol(self) -> str:
+        return self.tags.get("protocol", "unknown")
+
+    def set_metrics_by_protocol(self) -> None:
+        if self.protocol == "http":
+            self.metrics = PrometheusHTTPClientMetrics()
+        elif self.protocol == "thrift":
+            # self.metrics = PrometheusThriftClientMetrics()
+            logger.debug("PrometheusThriftClientMetrics not implemented.")
+        else:
+            logger.warning(
+                "No valid protocol set for Prometheus metric collection, metrics won't be collected. Expected 'http' or 'thrift' protocol. Actual protocol: %s",
+                self.protocol,
+            )
+        return
+
+    def on_start(self) -> None:
+        self.set_metrics_by_protocol()
+        if self.metrics is None:
+            logger.warning(
+                "No metrics set for Prometheus metric collection. Metrics will not be exported correctly."
+            )
+            return
+        self.start_time = time.perf_counter_ns()
+        self.metrics.active_requests_metric(self.tags).inc()
+
+    def on_incr_tag(self, key: str, delta: float) -> None:
+        pass
+
+    def on_finish(self, exc_info: Optional[_ExcInfo]) -> None:
+        if self.metrics is None:
+            logger.warning(
+                "No metrics set for Prometheus metric collection. Metrics will not be exported correctly."
+            )
+            return
+
+        self.tags["success"] = "true"
+        if exc_info is not None:
+            self.tags["exception_type"] = exc_info[1].__class__.__name__
+            self.tags["success"] = "false"
```

Comment:
Where is this used?

Reply:
We use this for Thrift (only), and it shares this code.